state/state_sqlite.go

```
@@ -0,0 +1,151 @@
+package state
+
+import (
+	"database/sql"
+	"errors"
+)
+
+// Verify that SyncState interface is implemented.
+var _ SyncState = (*sqliteSyncState)(nil)
+
+type sqliteSyncState struct {
+	db *sql.DB
+}
+
+func NewPersistentSyncState(db *sql.DB) *sqliteSyncState {
+	return &sqliteSyncState{db: db}
+}
+
+func (p *sqliteSyncState) Add(newState State) error {
+	var groupIDBytes []byte
+	if newState.GroupID != nil {
+		groupIDBytes = newState.GroupID[:]
+	}
+
+	_, err := p.db.Exec(`
+		INSERT INTO mvds_states 
+			(type, send_count, send_epoch, group_id, peer_id, message_id) 
+		VALUES 
+			(?, ?, ?, ?, ?, ?)`,
+		newState.Type,
+		newState.SendCount,
+		newState.SendEpoch,
+		groupIDBytes,
+		newState.PeerID[:],
+		newState.MessageID[:],
+	)
+	return err
+}
+
+func (p *sqliteSyncState) Remove(messageID MessageID, peerID PeerID) error {
+	result, err := p.db.Exec(
+		`DELETE FROM mvds_states WHERE message_id = ? AND peer_id = ?`,
+		messageID[:],
+		peerID[:],
+	)
+	if err != nil {
+		return err
+	}
+	if n, err := result.RowsAffected(); err != nil {
+		return err
+	} else if n == 0 {
+		return errors.New("state not found")
```

Are we sure we want to return an error when the state is not there? feels more like a warning (returning 0, and handled on the caller if necessary). If we do decide to return an error (I am fine with it), then it's worth having `ErrStateNotFound` or something so we can handle it on the caller, what do you reckon?

---

Totally. I missed it here but it should be an exported `error` instance.